✨ Use `ruff` for Linting

shaneahmed opened the pull request:
- Add checks for 
    - pydocstyle
    - flake8-future-annotations
    - flake8-errmsg
    - pygrep-hooks
    - flake8-use-pathlib
    - Ruff-specific rules
    - flake8-self
    - flake8-type-checking
    - tryceratops

shaneahmed commented on pyproject.toml:
```diff
@@ -51,3 +51,98 @@
+  "PYI",   # flake8-pyi
+  "RET",   # flake8-return
+  "RSE",   # flake8-raise
+  "RUF",   # Ruff-specific rules
```
```suggestion
  # "RUF",   # Ruff-specific rules
```

shaneahmed commented on tiatoolbox/tools/graph.py:
```diff
@@ -4,17 +4,19 @@
 from scipy.cluster import hierarchy
 from scipy.spatial import Delaunay, cKDTree
 
+if TYPE_CHECKING:  # pragma: no_cover
```
```suggestion
if TYPE_CHECKING:  # pragma: no cover
```

shaneahmed commented on tiatoolbox/tools/registration/wsi_registration.py:
```diff
@@ -16,16 +19,18 @@
-from tiatoolbox.wsicore.wsimeta import Resolution, Units
 from tiatoolbox.wsicore.wsireader import IntBounds, VirtualWSIReader, WSIReader
 
+if TYPE_CHECKING:  # pragma: no_cover
```
```suggestion
if TYPE_CHECKING:  # pragma: no cover
```

shaneahmed commented on tiatoolbox/wsicore/wsireader.py:
```diff
@@ -24,12 +24,14 @@
-from tiatoolbox.wsicore.metadata.ngff import Multiscales
 from tiatoolbox.wsicore.wsimeta import Resolution, Units, WSIMeta
 
+if TYPE_CHECKING:  # pragma: no_cover
```
```suggestion
if TYPE_CHECKING:  # pragma: no cover
```

shaneahmed commented:
@mostafajahanifar Please can you review `models`?

shaneahmed commented:
@measty Please can you review `tools`, `data`, `cli` and `visualization`?

shaneahmed commented:
@adamshephard Please can you review tests?

shaneahmed commented:
@John-P Please can you review `wsicore`, `utils` and root?

shaneahmed commented on .pre-commit-config.yaml:
```diff
@@ -75,26 +75,9 @@ repos:
-    rev: 6.0.0
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    # Ruff version.
+    rev: v0.0.276
```
```suggestion
    rev: v0.0.277
```

shaneahmed commented on tests/models/test_arch_vanilla.py:
```diff
@@ -36,13 +36,16 @@ def test_functional():
-    for backbone in backbones:
-        try:
+
+    backbone = "empty"
```
```suggestion
    # Dummy entry, will generate ValueError if "try" fails without running the loop.
    backbone = "empty"
```

shaneahmed commented on tests/test_tileserver.py:
```diff
@@ -148,7 +148,7 @@ def test_get_index(app) -> None:
 
 def test_create_with_dict(sample_svs):
-    """test initialising with layers dict"""
+    """Test initialising with layers dict."""
```
```suggestion
    """Test initializing with layers dict."""
```

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -590,31 +609,33 @@ def assert_dtype_int(input_var, message="Input must be integer."):
-        url (path): URL from where to download the data.
-        save_path (str): Location to unzip the data.
-        overwrite (bool): True to force overwriting of existing data, default=False
+        url: (path): URL from where to download the data.
+        save_path: (str): Location to unzip the data.
+        overwrite: (bool): True to force overwriting of existing data, default=False
```
```suggestion
        url (path): 
            URL from where to download the data.
        save_path (str): 
            Location to unzip the data.
        overwrite (bool): 
            True to force overwriting of existing data, default=False
```

shaneahmed commented on tiatoolbox/utils/magic.py:
```diff
@@ -35,16 +35,20 @@ def _normalize_binaryio(
+                raise FileNotFoundError(msg)
             return BytesIO()
-        return open(file, "rb")  # noqa: SIM115 -- intentional
+        return Path.open(file, "rb")  # -- intentional
```
```suggestion
        return Path.open(path, "rb")  # -- intentional
```

shaneahmed commented on tiatoolbox/models/engine/semantic_segmentor.py:
```diff
@@ -714,13 +748,16 @@ def _predict_one_wsi(
         """
-        cache_dir = f"{self._cache_dir}/{wsi_idx}/"
-        os.makedirs(cache_dir)
+        cache_dir = self._cache_dir / wsi_idx
```
```suggestion
        cache_dir = self._cache_dir / str(wsi_idx)
```

shaneahmed commented on tiatoolbox/models/engine/semantic_segmentor.py:
```diff
@@ -714,13 +748,16 @@ def _predict_one_wsi(
-        cache_dir = f"{self._cache_dir}/{wsi_idx}/"
-        os.makedirs(cache_dir)
+        cache_dir = self._cache_dir / wsi_idx
+        Path.mkdir(cache_dir, parents=True)
```
```suggestion
        cache_dir.mkdir(parents=True)
```

shaneahmed commented on tests/models/test_nucleus_instance_segmentor.py:
```diff
@@ -539,7 +542,7 @@ def test_cli_nucleus_instance_segment_ioconfig(remote_sample, tmp_path):
     }
 
-    with open(tmp_path.joinpath("config.yaml"), "w") as fptr:
+    with Path.open(tmp_path.joinpath("config.yaml"), "w") as fptr:
```
```suggestion
    with Path.open(tmp_path / "config.yaml", "w+") as fptr:
```

shaneahmed commented on tiatoolbox/models/engine/semantic_segmentor.py:
```diff
@@ -23,9 +21,13 @@
-from tiatoolbox.wsicore.wsimeta import Resolution, Units
 from tiatoolbox.wsicore.wsireader import VirtualWSIReader, WSIMeta, WSIReader
 
+if TYPE_CHECKING:
```
```suggestion
if TYPE_CHECKING:  # pragma: no cover
```

shaneahmed replied:
ignore type checking coverage

John-P commented on setup.py:
```diff
@@ -6,10 +6,10 @@
 
-with open("README.md") as readme_file:
+with Path.open(Path("README.md")) as readme_file:
     readme = readme_file.read()
 
-with open("HISTORY.md") as history_file:
+with Path.open(Path("HISTORY.md")) as history_file:
     history = history_file.read()
```
Are these Path.open changes suggestions from ruff?

shaneahmed replied:
ruff suggests using `Path.open()` which generates an error with `str`. I think it should be Path("HISTORY.md").open()

John-P commented on tiatoolbox/models/architecture/idars.py:
```diff
@@ -58,20 +60,20 @@ class IDaRS(CNNModel):
 
     @staticmethod
-    # noqa: E800
     def preproc(image: np.ndarray):
-        """Define preprocessing steps.
+        """Defines preprocessing steps.
```
I think it should be 'Define' not 'Defines'. Generally, function descriptions should be in the imperative case e.g. 'Compute', 'Solve...' examples at: https://docs.scipy.org/doc/scipy/reference/linalg.html.

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/commit/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/models/architecture/mapde.py:
```diff
@@ -77,7 +77,8 @@ def __init__(
         num_classes: int = 1,
-    ):
+    ) -> None:
+        """Initializes :class:`MapDe`."""
```
```suggestion
        """Initialize :class:`MapDe`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/models/architecture/nuclick.py:
```diff
@@ -60,7 +60,8 @@ def __init__(
         do_batchnorm: bool = True,
-    ):
+    ) -> None:
+        """Initializes :class:`ConvBnRelu`."""
```
```suggestion
        """Initialize :class:`ConvBnRelu`."""
```

John-P commented on tiatoolbox/models/architecture/nuclick.py:
```diff
@@ -119,7 +120,7 @@ def get_block(
                 Size of stride in the convolution layer.
             use_bias (bool):
                 Whether to use bias in the convolution layer.
-            dilation_rates (list):
+            dilation_rate (list):
```
I generally think arguments which must be sequences such as a numpy array, tuple, list etc. should be plural to indicate that it is not a single value / scalar.

John-P commented on tiatoolbox/models/architecture/nuclick.py:
```diff
@@ -188,7 +189,8 @@ def __init__(
         use_bias: bool = False,
-    ):
+    ) -> None:
+        """Initializes :class:`MultiscaleConvBlock`."""
```
```suggestion
        """Initialize :class:`MultiscaleConvBlock`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/models/architecture/nuclick.py:
```diff
@@ -285,7 +288,8 @@ def __init__(
         dilation_rate: Union[Tuple[int, int], np.ndarray] = (1, 1),
-    ):
+    ) -> None:
+        """Initializes :class:`ResidualConv`."""
```
```suggestion
        """Initialize :class:`ResidualConv`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/models/architecture/nuclick.py:
```diff
@@ -353,7 +357,8 @@ class NuClick(ModelABC):
 
-    def __init__(self, num_input_channels: int, num_output_channels: int):
+    def __init__(self, num_input_channels: int, num_output_channels: int) -> None:
+        """Initializes :class:`NuClick`."""
```
```suggestion
        """Initialize :class:`NuClick`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/models/architecture/vanilla.py:
```diff
@@ -193,7 +200,8 @@ class CNNBackbone(ModelABC):
 
-    def __init__(self, backbone):
+    def __init__(self, backbone) -> None:
+        """Initializes :class:`CNNBackbone`."""
```
```suggestion
        """Initialize :class:`CNNBackbone`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/models/dataset/classification.py:
```diff
@@ -103,6 +106,7 @@ def __init__(self, inputs, labels=None):
         self._check_input_integrity(mode="patch")
 
     def __getitem__(self, idx):
+        """Defines the behaviour when an item is accessed."""
```
```suggestion
        """Get an item from the dataset."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/a24a4918731addef5dc1da96c9fe6e396b122a37

John-P commented on tiatoolbox/models/dataset/classification.py:
```diff
@@ -323,6 +332,7 @@ def __init__(
         self._check_input_integrity(mode="wsi")
 
     def __getitem__(self, idx):
+        """Defines the behaviour when an item is accessed."""
```
```suggestion
        """Get an item from the dataset."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/a24a4918731addef5dc1da96c9fe6e396b122a37

John-P commented on tiatoolbox/models/dataset/dataset_abc.py:
```diff
@@ -13,7 +13,8 @@ class PatchDatasetABC(ABC, torch.utils.data.Dataset):
         self,
-    ):
+    ) -> None:
+        """Initializes :class:`PatchDatasetABC`."""
```
```suggestion
        """Initialize :class:`PatchDatasetABC`."""
```

John-P commented on tiatoolbox/models/dataset/dataset_abc.py:
```diff
@@ -143,11 +145,14 @@ def preproc_func(self, func):
 
-    def __len__(self):
+    def __len__(self) -> int:
+        """Returns the length of the instance attributes."""
```
```suggestion
        """Return the length of the instance attributes."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/2db02db7a6949e400ee54fe03a5dfa0021412f44

John-P commented on tiatoolbox/models/dataset/dataset_abc.py:
```diff
@@ -143,11 +145,14 @@ def preproc_func(self, func):
 
     @abstractmethod
     def __getitem__(self, idx):
+        """Defines the behaviour when an item is accessed."""
```
```suggestion
        """Get an item from the dataset."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/a24a4918731addef5dc1da96c9fe6e396b122a37

John-P commented on tiatoolbox/models/dataset/info.py:
```diff
@@ -70,7 +73,8 @@ class name as a string.
         save_dir_path=None,
-    ):
+    ) -> None:
+        """Initializes :class:`KatherPatchDataset`."""
```
```suggestion
        """Initialize :class:`KatherPatchDataset`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tests/test_annotation_stores.py:
```diff
@@ -2091,6 +2116,7 @@ def test_nquery_overlapping_grid_poly_poly(store_cls):
 
     @staticmethod
     def test_invalid_mode_type(store_cls):
+        """Tests invalid mode type for AnnotationStore."""
```
```suggestion
        """Test invalid mode type for AnnotationStore."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/d5bc38668bd092f49acc621c6086ff35776bb23f

John-P commented on tests/test_annotation_stores.py:
```diff
@@ -2115,6 +2142,7 @@ def test_invalid_mode_format(store_cls):
 
     @staticmethod
     def test_invalid_mode(store_cls):
+        """Tests invalid mode for AnnotationStore."""
```
```suggestion
        """Check unsupported mode raises ValueError."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/d5bc38668bd092f49acc621c6086ff35776bb23f

John-P commented on tests/test_annotation_stores.py:
```diff
@@ -2103,6 +2129,7 @@ def test_invalid_mode_type(store_cls):
 
     @staticmethod
     def test_invalid_mode_format(store_cls):
+        """Tests invalid mode format for AnnotationStore."""
```
```suggestion
        """Check invalid mode string format raises ValueError."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/d5bc38668bd092f49acc621c6086ff35776bb23f

John-P commented on tests/test_wsi_registration.py:
```diff
@@ -20,24 +21,27 @@
 
 def test_extract_features(dfbr_features):
     """Test for CNN based feature extraction function."""
-    df = DFBRegister()
```
This could probably have a more descriptive name

shaneahmed replied:
Added a description for dfbr in comments.

John-P commented on tests/test_wsireader.py:
```diff
@@ -1078,7 +1093,7 @@ def test_virtual_wsi_reader_invalid_mode(source_image):
 
 
 def test_virtual_wsi_reader_read_bounds(source_image):
```
We should split this into smaller test. It is trying to do too many things in one.

shaneahmed replied:
In a separate PR I would suggest.

John-P commented on tiatoolbox/models/engine/patch_predictor.py:
```diff
@@ -28,7 +27,8 @@ def __init__(
         **kwargs,
-    ):
+    ) -> None:
+        """Initializes :class:`IOPatchPredictorConfig`."""
```
```suggestion
        """Initialize :class:`IOPatchPredictorConfig`."""
```

John-P commented on tiatoolbox/models/engine/semantic_segmentor.py:
```diff
@@ -311,11 +327,12 @@ class WSIStreamDataset(torch_data.Dataset):
         mode="wsi",
-    ):
+    ) -> None:
+        """Initializes :class:`WSIStreamDataset`."""
```
```suggestion
        """Initialize :class:`WSIStreamDataset`."""
```

John-P commented on tiatoolbox/models/models_abc.py:
```diff
@@ -15,24 +15,26 @@ class IOConfigABC(ABC):
 
-    def __init__(self):
+    def __init__(self) -> None:
+        """Initializes Abstract class ModelABC."""
```
```suggestion
        """Initialize Abstract class ModelABC."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/tools/patchextraction.py:
```diff
@@ -142,24 +146,31 @@ def __init__(
         return self.locations_df.shape[0] if self.locations_df is not None else 0
 
     def __next__(self):
+        """Returns the next item for the iteration."""
```
```suggestion
        """Return the next item for the iteration."""
```

John-P commented on tiatoolbox/tools/patchextraction.py:
```diff
@@ -142,24 +146,31 @@ def __init__(
 
-    def __len__(self):
+    def __len__(self) -> int:
+        """Returns the length of the instance attributes."""
```
```suggestion
        """Return the length of the instance attributes."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/2db02db7a6949e400ee54fe03a5dfa0021412f44

John-P commented on tiatoolbox/tools/patchextraction.py:
```diff
@@ -142,24 +146,31 @@ def __init__(
         self.within_bound = within_bound
 
     def __iter__(self):
+        """Returns an iterator for the given object."""
```
```suggestion
        """Return an iterator for the given object."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/2db02db7a6949e400ee54fe03a5dfa0021412f44

John-P commented on tiatoolbox/tools/stainaugment.py:
```diff
@@ -101,13 +102,14 @@ class StainAugmentor(ImageOnlyTransform):
         p=0.5,
-    ) -> np.ndarray:
+    ) -> None:
+        """Initializes :class:`StainAugmentor`."""
```
```suggestion
        """Initialize :class:`StainAugmentor`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/tools/stainextract.py:
```diff
@@ -79,10 +79,12 @@ class CustomExtractor:
 
-    def __init__(self, stain_matrix):
+    def __init__(self, stain_matrix) -> None:
+        """Initializes :class:`CustomExtractor`."""
```
```suggestion
        """Initialize :class:`CustomExtractor`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/tools/stainextract.py:
```diff
@@ -158,7 +161,8 @@ class MacenkoExtractor:
 
-    def __init__(self, luminosity_threshold=0.8, angular_percentile=99):
+    def __init__(self, luminosity_threshold=0.8, angular_percentile=99) -> None:
+        """Initializes :class:`MacenkoExtractor`."""
```
```suggestion
        """Initialize :class:`MacenkoExtractor`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/tools/stainextract.py:
```diff
@@ -241,7 +246,8 @@ class VahadaneExtractor:
 
-    def __init__(self, luminosity_threshold=0.8, regularizer=0.1):
+    def __init__(self, luminosity_threshold=0.8, regularizer=0.1) -> None:
+        """Initializes :class:`VahadaneExtractor`."""
```
```suggestion
        """Initialize :class:`VahadaneExtractor`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/tools/stainnorm.py:
```diff
@@ -149,7 +154,8 @@ class RuifrokNormalizer(StainNormalizer):
 
-    def __init__(self):
+    def __init__(self) -> None:
+        """Initializes :class:`RuifrokNormalizer`."""
```
```suggestion
        """Initialize :class:`RuifrokNormalizer`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/tools/tissuemask.py:
```diff
@@ -17,6 +17,7 @@ class TissueMasker(ABC):
     """
 
     def __init__(self) -> None:
+        """Initializes :class:`TissueMasker`."""
```
```suggestion
        """Initialize :class:`TissueMasker`."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P commented on tiatoolbox/utils/metrics.py:
```diff
@@ -69,7 +69,9 @@ def f1_detection(true, pred, radius):
 
 def dice(gt_mask, pred_mask):
-    r"""This function computes `Sørensen–Dice coefficient
+    r"""Computes Sørensen-Dice coefficient.
```
```suggestion
    r"""Compute the Sørensen-Dice coefficient.
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/6ae8838cdb92759eb927005fd350f505d95fba9f

John-P commented on tiatoolbox/utils/misc.py:
```diff
@@ -342,8 +353,9 @@ def mpp2objective_power(mpp):
 def contrast_enhancer(img, low_p=2, high_p=98):
-    """Enhancing contrast of the input image using intensity adjustment.
-       This method uses both image low and high percentiles.
+    """Enhances contrast of the input image using intensity adjustment.
```
```suggestion
    """Enhance contrast of the input image using intensity adjustment.
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/7deae026e07b52b86558f194e9231b630c756b33

John-P commented on tiatoolbox/wsicore/wsimeta.py:
```diff
@@ -103,7 +103,8 @@ def __init__(
         raw: Optional[Mapping[str, str]] = None,
-    ):
+    ) -> None:
+        """Initializas WSIMeta."""
```
```suggestion
        """Initialize WSIMeta."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/c2e1cd4ccfd6aaeb32664f23d4f6634cddb90090

John-P reviewed (commented):
This PR is quite extensive! It seems that the majority of it involves organizing and cleaning up, likely in response to ruff suggestions. I provided a few notes on naming conventions and documentation strings.

shaneahmed commented on requirements/requirements.txt:
```diff
@@ -26,7 +26,7 @@ SimpleITK>=2.2.1
 sphinx>=5.3.0
 tifffile>=2022.10.10
 torch>=2.0.0
-torchvision>=0.14.1
```
Good catch. No it was not intentional.

shaneahmed replied:
Actually it is added on the next line with >0.15.0. Version 0.14.x doesnot support WeightsEnum

shaneahmed commented on tiatoolbox/visualization/tileserver.py:
```diff
@@ -247,6 +251,8 @@ def zoomify(
             layer (str):
                 The layer name.
+            session_id (str):
+                Session ID.
```
@measty Please can you check this?

shaneahmed replied:
```suggestion
                Session ID. Unique ID to disambiguate requests from different sessions.
```

shaneahmed commented on .pre-commit-config.yaml:
```diff
@@ -75,26 +75,9 @@ repos:
-    rev: 6.0.0
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    # Ruff version.
+    rev: v0.0.277
```
```suggestion
    rev: v0.0.278
```

shaneahmed commented:
> @measty Please can you review `tools`, `data`, `cli`, `annotation` and `visualization`?

@measty  copying again as the previous comment is lost in the commits.

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2143,22 +2218,26 @@ def _initialize_query_string_parameters(
         no_constraints_ok: bool = False,
         index_warning: bool = False,
-        min_area: Optional[float] = None,
+        min_area: float | None | None = None,
```
None is repeated

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2205,14 +2284,18 @@ def _query(
             callable_columns = columns
         if geometry_predicate not in self._geometry_predicate_names:
+            msg = (
+                f"Invalid geometry predicate.Allowed values are: "
```
no f-string, add space

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2383,10 +2469,10 @@ def query(
-        min_area: Optional[float] = None,
-    ) -> Dict[str, Tuple[float, float, float, float]]:
+        geometry: QueryGeometry | None = None,
+        where: str | bytes | Callable[[Geometry, dict[str, Any]], bool] | None = None,
```
we have some custom types we can use here, I think we could replace this with Predicate | None

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2585,9 +2671,9 @@ def _handle_str_pquery(
-        where: Union[str, bytes, Callable],
-    ) -> Tuple[bool, bool, bool]:
+        select: str | bytes | Callable,
+        where: str | bytes | Callable,
```
Again we could use custom defined types here, Select and Predicate

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2610,8 +2696,8 @@ def _kind_of_pquery(
-        select: Union[str, bytes, Callable],
-        where: Union[str, bytes, Callable],
+        select: str | bytes | Callable,
+        where: str | bytes | Callable,
```
Select and Predicate custom types

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2773,17 +2868,20 @@ def pquery(
         return result
 
     def __len__(self) -> int:
+        """Return the length of the instance attributes."""
```
More descriptive? 'Return number of annotations in the store.'

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -2996,6 +3099,7 @@ def remove_many(self, keys: Iterable[str]) -> None:
             self.con.commit()
 
     def __setitem__(self, key: str, annotation: Annotation) -> None:
+        """Implements a method to assign a value to and item."""
```
typo, should be: to an item

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -3273,26 +3387,30 @@ def remove(self, key: str) -> None:
         del self._rows[key]
 
     def __getitem__(self, key: str) -> Annotation:
+        """Get an item from the dataset."""
```
Get an item from the store.

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty commented on tiatoolbox/annotation/storage.py:
```diff
@@ -3273,26 +3387,30 @@ def remove(self, key: str) -> None:
         return self._rows[key]["annotation"]
 
     def __setitem__(self, key: str, annotation: Annotation) -> None:
+        """Implements a method to assign a value to and item."""
```
an, not and

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

measty reviewed (commented):
I've reviewed the parts assigned to me and left a few minor comments. In general the PR looks good.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1447,10 +1491,10 @@ def transform_geometry(geom):
         ]
 
-        print(f"added {len(annotations)} annotations")
+        logger.info("Added %d annotations.", len(annotations))
```
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -803,15 +826,18 @@ def iquery(
         """
         if geometry_predicate not in self._geometry_predicate_names:
+            msg = (
+                f"Invalid geometry predicate.Allowed values are: "
```
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -671,17 +683,21 @@ def query(
+            raise ValueError(msg)
         if geometry_predicate not in self._geometry_predicate_names:
+            msg = (
+                f"Invalid geometry predicate.Allowed values are: "
```
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

shaneahmed commented on tiatoolbox/annotation/dsl.py:
```diff
@@ -79,88 +83,116 @@ class SQLExpression:
         return SQLTriplet(self, operator.ne, other)
 
     def __neg__(self):
+        """Define how the object is compared for equality (not equal to)."""
```
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

shaneahmed commented on tiatoolbox/tools/patchextraction.py:
```diff
@@ -142,24 +152,31 @@ def __init__(
 
-    def __len__(self):
+    def __len__(self) -> int:
+        """Return the length of the instance attributes."""
```
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/991322d859c0c22e7c7a1d713362944aa4317198

shaneahmed commented:
Thanks @measty I have addressed your comments.

Please can you add comments on line 255 and 265 in the tile_server docstrings so I can resolve these. I have added a link below.
https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/files/254b7555cb9c08cc7fe68b5d264baeb817c21beb..c0053681c8cae89ccf2c454286007191fd4402d4#diff-85022a59fadc36909d873efb715225154f0f0cda174af2235f80ac85bf42429a

shaneahmed commented:
> Session ID.

@John-P

shaneahmed commented:
> @John-P Please can you review `wsicore`, `utils` and root?

@John-P

shaneahmed commented on tiatoolbox/visualization/tileserver.py:
```diff
@@ -255,6 +261,8 @@ def zoomify(
             y (int):
                 The y coordinate.
+            res (int):
+                Display Resolution.
```
```suggestion
                Resolution to save the tiles at. Helps to specify high resolution tiles. Valid options are 1 and 2.
```

John-P commented on tiatoolbox/utils/env_detection.py:
```diff
@@ -23,13 +24,15 @@
 from tiatoolbox import logger
 
+if TYPE_CHECKING:  # pragma: no cover
+    from numbers import Number
```
We might want to swap from Number to a custom type e.g. Union[int, float]. See this thread: https://stackoverflow.com/questions/60616802/how-to-type-hint-a-generic-numeric-type-in-python, which suggests using:
```python
from typing import SupportsFloat as Numeric
```
Some funky things can happen with Number e.g. because it includes complex numbers.

shaneahmed replied:
https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/bb24fccd10322177197c25b2192531eb94e9a6b4

John-P commented on tiatoolbox/__init__.py:
```diff
@@ -46,7 +50,15 @@
+    """
+
+    def filter(self, record):  # noqa: A003
+        """Filters input record."""
```
```suggestion
        """Filter records."""
```

shaneahmed replied:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/625/commits/bb24fccd10322177197c25b2192531eb94e9a6b4

John-P reviewed (commented):
Added a couple of comments

shaneahmed commented on tiatoolbox/utils/env_detection.py:
```diff
@@ -249,7 +249,9 @@ def has_network(
-def pixman_versions() -> list[tuple[int, ...]]:  # noqa: CCR001
+
+def pixman_versions() -> list[tuple[int, ...]]:
+=======
```
```suggestion
```
